### ISSUE TITLE
[ICU BUG] Incorrect behavior of look-behind assertion with stri_match/stri_extract/stri_split_regex

### ISSUE THREAD

**Issue by kohske** [gagolews/stringi#147]
```
> stri_match("åfoo", regex = "(?<=å)foo") # OK
     [,1] 
[1,] "foo"
> stri_match("aåfoo", regex = "(?<=aå)foo") # NG
     [,1]
[1,] NA  
> stri_match("aafoo", regex = "(?<=aa)foo") # OK
     [,1] 
[1,] "foo"
> 
> stri_match("åfoo", regex = "(?<!å)foo") # OK
     [,1]
[1,] NA  
> stri_match("aåfoo", regex = "(?<!aå)foo") # NG
     [,1] 
[1,] "foo"
> stri_match("aafoo", regex = "(?<!aa)foo") # OK
     [,1]
[1,] NA  
> 
> stri_detect("åfoo", regex = "(?=!å)foo") # OK
[1] FALSE
> stri_detect("aåfoo", regex = "(?=!aå)foo") # OK
[1] FALSE
> stri_detect("aafoo", regex = "(?=!aa)foo") # OK
[1] FALSE
> stri_detect("åfoo", regex = "(?<!å)foo") # OK
[1] FALSE
> stri_detect("aåfoo", regex = "(?<!aå)foo") # OK
[1] FALSE
> stri_detect("aafoo", regex = "(?<!aa)foo") # OK
[1] FALSE
> 
```

It fails when look-behind includes >=2 characters and the last character in the look-behind is non-ascii.

``` r
> sessionInfo()
R version 3.1.1 (2014-07-10)
Platform: x86_64-apple-darwin13.1.0 (64-bit)

locale:
[1] ja_JP.UTF-8/ja_JP.UTF-8/ja_JP.UTF-8/C/ja_JP.UTF-8/ja_JP.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] stringi_0.4-1 stringr_0.6.2

loaded via a namespace (and not attached):
[1] tools_3.1.1
```

**Comment by gagolews** [comment]
@kohske 
seems like a bug in the ICU library. I will check if everything's ok with icu 55milestone1. If not, I'll file a bug report.

**Comment by gagolews** [comment]
hmm... but stri_detect & str_count work OK (they base on regex UTF16 interface)

**Comment by kohske** [comment]
Although I have not inspected deeply, this may not be a bug in ICU,
At least I confirmed that the result of ICU's `RegexMatche::find()` is correct in all the cases above.

**Comment by gagolews** [comment]
But in this case I use the `UText` interface

**Comment by gagolews** [comment]
Minimal reproducible example:

```
#include <unicode/regex.h>
#include <cstdio>
#include <cstring>

/*
   g++ -std=c++11 icu_utext_regex_bug.cpp -licui18n -licuuc -licudata  && ./a.out
*/

int main() {
   UErrorCode        status    = U_ZERO_ERROR;
   RegexMatcher *matcher = new RegexMatcher("(?<=aą)foo", 0, status);
   if (U_FAILURE(status)) {
      return 1;
   }

   
   const char* stringToTest = "**aąfoo**";
   UText* str_text = utext_openUTF8(NULL, stringToTest, strlen(stringToTest), &status);
   matcher->reset(str_text);

   if (matcher->find()) {
      // We found a match.
      int startOfMatch = matcher->start(status);   // string index of start of match.
      printf("match at = %d\n", startOfMatch);
   }
   else
      printf("no match\n");
      
   return 0;
}
```

**Comment by gagolews** [comment]
The same if we use `UText` for `RegexMatcher`

```
#include <unicode/regex.h>
#include <cstdio>
#include <cstring>

/*
   g++ -std=c++11 icu_utext_regex_bug.cpp -licui18n -licuuc -licudata  && ./a.out
*/

int main() {
   UErrorCode        status    = U_ZERO_ERROR;
   const char* m = "(?<=ąbcą)foą";
   UText* str_m = utext_openUTF8(NULL, m, strlen(m), &status);
   if (U_FAILURE(status)) {
      printf("1error %s!\n", u_errorName(status));
      return 1;
   }
   
   status    = U_ZERO_ERROR;
   RegexMatcher *matcher = new RegexMatcher(str_m, 0, status);
   if (U_FAILURE(status)) {
      printf("2error %s!\n", u_errorName(status));
      return 1;
   }

   
   status    = U_ZERO_ERROR;
   const char* stringToTest = "**ąbcąfoą**";
   UText* str_text = utext_openUTF8(NULL, stringToTest, strlen(stringToTest), &status);
   if (U_FAILURE(status)) {
      printf("3error %s!\n", u_errorName(status));
      return 1;
   }
   
   
   matcher->reset(str_text);

   if (matcher->find()) {
      // We found a match.
      int startOfMatch = matcher->start(status);   // string index of start of match.
      printf("match at = %d\n", startOfMatch);
   }
   else
      printf("no match\n");
      
   return 0;
}
```


---> `no match`

**Comment by kohske** [comment]
Thanks, as you image I did test with the plain char.
I also reproduced the example above.
Obviously this is a bug in ICU.

**Comment by gagolews** [comment]
@kohske which `ICU` version do you use?

**Comment by kohske** [comment]
perhaps 54.1. I installed it via homebrew.

**Comment by gagolews** [comment]
```
printf("U_ICU_VERSION=%s\n", U_ICU_VERSION);
```

**Comment by kohske** [comment]
yes exactly `U_ICU_VERSION=54.1`

**Comment by gagolews** [comment]
you're fast :)

Ok, just compiler ICU 54.1 -> confirmed

```
g++ -std=c++11 icu_utext_regex_bug.cpp -licui18n -licuuc -licudata  -I/opt/icu4c-54_1/include/ -L/opt/icu4c-54_1/lib/  && LD_LIBRARY_PATH="/opt/icu4c-54_1/lib/" ./a.out
U_ICU_VERSION=54.1
no match
```

filing a bug repo

**Comment by kohske** [comment]
Thanks!!
Anyway, thanks for the great package.

**Comment by gagolews** [comment]
http://bugs.icu-project.org/trac/ticket/11554

**Comment by gagolews** [comment]
OK, ICU guys confirmed it's a bug.

UTF-16 interface also doesn't work for 2-byte UTF-16

```
> stri_count_regex('***a\U0010000foo**', 'a\U0010000foo')
[1] 1
> stri_count_regex('***a\U0010000foo**', '(?<=a\U0010000)foo')
[1] 0
```

**Comment by gagolews** [comment]
Funny, `str_detect` doesn't induce this error. It operates on a UTF-16 string. On the other hand, `stri_extract` plays with `UText` on a UTF-8 string.

```
> expect_identical(stri_detect_regex("aafoo - ICU BUG TEST", "(?<=aa)foo"), TRUE)
> expect_identical(stri_detect_regex("aąfoo - ICU BUG TEST", "(?<=aą)foo"), TRUE)
> expect_identical(stri_extract_all_regex("aafoo - ICU BUG TEST", "(?<=aa)foo")[[1]], "foo")
> expect_identical(stri_extract_all_regex("aąfoo - ICU BUG TEST", "(?<=aą)foo")[[1]], "foo")
 Error: stri_extract_all_regex("aąfoo - ICU BUG TEST", "(?<=aą)foo")[[1]] is not identical to "foo". Differences: 
'is.NA' value mismatch: 1 in current 0 in target 
```

**Comment by drimacus182** [comment]
@gagolews It looks like this bug is fixed in ICU 57.1 version. Is there a chance to make stringi use this new version of ICU?

**Comment by gagolews** [comment]
Yep, this has been fixed, thanks for the info. I'll update the ICU bundle to v.58 as soon as I'll have some time for that.
